### ISSUE TITLE
regression when window switching/rescaling

### ISSUE THREAD

**Issue by suisseWalter** [xournalpp/xournalpp#3224]
(Please complete the following information, and then delete this line)

**Affects versions :**
 - OS: Ubuntu 20.04 LTS
 - (Linux only) Desktop environment: I3-X11
 - Which version of libgtk do you use 3.22.30 /(3.24.20 when built from source)
 - Version of Xournal++: [e.g. 1.0.16, specific commit, specific branch]
     1a60605e(nightly build 11 july 2021)
 - Installation method: appimage

**Describe the bug**
using I3 as my WM I have noticed some massive lags with switching to and from the latest version of nightly. 
It takes about a second to load the window when switching to it. the same goes if it's re-scaled(closing another window beside it).
I can also produce it by force-closing the about menu(clsing it by keyboard comand, not be close button), but not while opening it. 
note moving it around the same workspace(without changing the size) works without a problem. 

I experience similar issues while switching to and from graphics heavy games, that aren't really made for re-scaling their windows.

this behavior was introduced between nightly 20210615 and the latest one(20210711)

**To Reproduce**
1) open latest nightly in I3
2a) remove windows on the same workspace.
2b) open the about menu and then close it again.
2c) switch to the workspace xournalpp is on from a different one.
3) see xournalpp behaving slowly. and leaving black bars for a couple of seconds.

**Expected behavior**
Xournalpp should immediately rescale windows and update it's content. 
the corect behaviour is already happening in stable as well as in nightly 20210615(and before)

**Screenshots of Problem**

**Additional context**
I would be happy to try all the nightly build between 6 of june and today to find the one that introduced the problem, should that be needed.

**Comment by rolandlo** [comment]
@suisseWalter We haven't changed much which might explain the regression you observe from June 15, 2021 and July 11, 2021. Commit 07f2aa685db21540b46b1dee3652a2fea7e8e821 is the only one which comes to mind on first sight, and even that commit doesn't look likely to me. It is possible that some dependency update created the regression. You would find out if you build from source as described in the [wiki](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md). Let me know if you need help with it.

Are you sure you can exclude a system update on your side as the problem?

**Comment by suisseWalter** [comment]
I can have both appimages running in parallel and observe it(or just the new one). so there shouldn't be a problem with dependencies. I can try to build from source. did it in the past so should be possible. 
I was also looking at 6354a33 but is glade only used for creating files and not part if the final package?

**Comment by rolandlo** [comment]
The glade files are part of the package, but they only define the graphical user interface. They should not cause any performance difference. To extract them you can run the AppImage with the `--appimage-extract` option. You can also check which libraries got updated.

**Comment by suisseWalter** [comment]
I was able to build from source. and I have the same problem, but it seams slightly less pronounced. I also noticed that if I close the about menu with the proper close button it works correctly. 
I also noticed that it only happens if I had the window selected the last time I was on the workspace. and how long it takes until it properly displays varies. bellow is a short video of it happening. with all three versions running. 

https://user-images.githubusercontent.com/42143099/125266319-0352d080-e306-11eb-824a-d9d8b33eae3a.mp4

are the old nightly builds somewhere available or do I have to build everything from source to find the commit that causes this?

**Comment by suisseWalter** [comment]
What I've now also noticed, is that the menu bar takes even longer to be activated. this can also be seen in the video. when one looks closely at the color of the text on the menu bar, it changes from gray to black a lot after initially loads the page. this as well only happens in the newer versions. this isn't really a problem. as no one needs to be able to switch to the window and immediately enter the settings or something like that. but I would expect it to be caused by the same underlying problem.

**Comment by rolandlo** [comment]
> are the old nightly builds somewhere available or do I have to build everything from source to find the commit that causes this?

The last few nightly builds  are avaiable on [Azure Pipelines](https://dev.azure.com/xournalpp/xournalpp/_build?definitionId=3), but I think only devs have access to it.

**Comment by suisseWalter** [comment]
jep. they aren't available for me. 
So I will build them from source. and binary search myself through the dates. that should be quite fast. 
how do I download the repository up to a certain commit?

**Comment by rolandlo** [comment]
just clone the repository as usual and then type 
`git checkout <hash>`
where `hash` is the hash-code of the commit after which you want to compile Xournal++. You can copy the `hash codes` from the [comit list](https://github.com/xournalpp/xournalpp/commits/master)
For instance 1a60605e7ccfe1d46162355825dd19f27d931db4 is currently the latest `hash code`

**Comment by suisseWalter** [comment]
After search through different commits I can conclude that the problem is caused by edfaf851. which makes sense. as this commit works a lot with the interface and rewrites a lot of code. Also I overlooked it the first time because it was commited on the 15th. and my last working version was from the 15th. but from what I understand azure builds the stuff at midnight and marks it for the day it releases. not the day the last commit is from. 
how should I procede from here? should I just try to split edfaf851 into different parts and try to figure out this way what exactly causes it?
also I noticed that the "Use available Stock Icons" doesn't work (will open a new issue about it) it will always use the new ones, I assume with Stock it's referring to the stock ones of the OS.?. because I thought, maybe the new icons just take a long time to load? but without this one working I can't really check.

**Comment by rolandlo** [comment]
@suisseWalter Are you sure the "Use available Stock Icons" doesn't work? Only some icons get exchanged (as there are not many stock icons from the OS available), see #3161. It worked on all systems I have tried. If that is indeed the case, you would get a hint where to look for. Otherwise I would first look at the changes in `XournalMain.cpp`. Older Gtk versions seem to have the load order of the icons switched somehow, which is both apparent in the AppImage (that is compiled on Ubuntu 18.04) and the .deb package built for Ubuntu 18.04.

**Comment by suisseWalter** [comment]
"Use available Stock Icons" works. I had the same issue as #3161 . I will start to take a look at `XournalMain.cpp`. but that might take a couple of days until I get the time to do it. 
I've looked through the commit a bit. and I've noticed that most of the svg's have increased by about 10x in size. a extreme example would be the `fullscreen.svg` which goes from 40 lines to 10'000 lines. 
I can confirm that this is causing the problem. because if I switch the toolbar over to "empty toolbar" I don't have this problem. and performance is the same as in the old version. 
I'm not sure how exactly the graphics stuff works. but I would think that it's reasonable that it might rerender the entire picture if I switch to a workspace. and this would probably lead to parsing all of these files again. which now would take ages. this also explains why it sometimes varies with how long it takes to load. because it probably depends on what stuff is still in ram/gram.

So now the big question is why are these files so massive. and why do they varie so massively? what are they exactly for?

**Comment by rolandlo** [comment]
Hmm, we really didn't look into the file size aspect, since we didn't experience any performance issues. But you are right that the files could probably a lot smaller. Using [svgo](https://github.com/svg/svgo) I got them in 2 iterations from 13mb to 375kb without any visible quality loss.

**Comment by suisseWalter** [comment]
I can confirm that scaling them down works. but it's still not perfect, there is still some lag. I'm wondering if there is some config option in gtk3 that would stop it from happening. because it's only the icons that cause the problem. I can have a page loaded that has strokes worth 40mb on it and this doesn't change how long it takes to load. 
but I would consider the problem fixed for now.

**Comment by rolandlo** [comment]
> I can confirm that scaling them down works. but it's still not perfect, there is still some lag. 

The same can be observed on MacOS BigSur. 

> I'm wondering if there is some config option in gtk3 that would stop it from happening. because it's only the icons that cause the problem. 

I don't know about that. Clearly it seems that the icons are re-rendered (and maybe re-loaded) much more often than it needs to be.
